File: python/doublylinked_list.py
class Node:
  def __init__(self, data):
    self.data = data
    self.next = None
    self.prev = None


class DoublyLinkedList:
  def __init__(self):
    self.head = None

  def insert_at_beginning(self, data):
    new_node = Node(data)
    new_node.next = self.head
    if self.head is not None:
      self.head.prev = new_node
    self.head = new_node


  def print_list(self):
    temp = self.head
    while temp:
      print(temp.data, end=" <-> ")
      temp = temp.next
    print("none")

# example usage
if __name__ == "__main__":
  dl1 = DoublyLinkedList()
  dl1.insert_at_beginning(10)
  dl1.insert_at_beginning(20)
  dl1.insert_at_beginning(30)

  print("doubly linked list: ")
  dl1.print_list()

File: c/doublyLinkedList.c
#include <stdio.h>
#include <stdlib.h>

// define node structure
struct Node {
  int data;
  struct Node* next;
  struct Node* prev;

};

// function to create a new node
struct Node* createNode(int data) {
  struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
  newNode->data = data;
  newNode->next = NULL:
  newNode->prev = NULL;
  return newNode;
}

// function to insert node at beginning
void insertAtBeginning(struct Node** head, int data) {
  struct Node* newNode = createNode(data);
  newNode->next = *head;
  if(*head != NULL) {
    (*head)->prev = newNode;
  }
  *head = newNode;
}

// function to print linked list 
void printlist(struct Node* head) {
  struct Node* temp = head;
  while(temp != NULL) {
    printf("%d <-> ", temp-> data);
    temp = temp->next;
  }
  printf("NULL\n");
}

// main function to test doubly linked list 
int main() {
  struct Node* head = NULL;

  insertAtBeginning(&head, 10);
  insertAtBeginning(&head, 20);
  insertAtBeginning(&head, 30);

  printf("doubly linked list; ");
  printList(head);
  return 0;
}

